The goal of this project is to understand the overall sales performance of the company.

Want to understand the following:

- Among all product lines, which one has the highest sales?
- Which year recorded the highest sales?
- Which month experienced the highest sales?
- Which country achieved the highest product sales?
- Who stands out as the top customer?
- What products are usually sold together?
  

Let's start with the data exploration.

In [1]:
%load_ext sql

In [2]:
# Connecting to the MySQL database
%sql mysql+mysqlconnector://root:0808Kari!@localhost/test_db

In [121]:
# Loading my SALES data
%sql SELECT * FROM sales_data LIMIT 10; 

 * mysql+mysqlconnector://root:***@localhost/test_db
10 rows affected.


[('10107', '30', '95.7', '2', '2871', '2003-02-24 00:00:00', 'Shipped', '1', '2', '2003', 'Motorcycles', '95', 'S10_1678', 'Land of Toys Inc.', '2125557818', '897 Long Airport Avenue', '', 'NYC', 'NY', '10022', 'USA', 'NA', 'Yu', 'Kwai', 'Small'),
 ('10121', '34', '81.35', '5', '2765.9', '2003-05-07 00:00:00', 'Shipped', '2', '5', '2003', 'Motorcycles', '95', 'S10_1678', 'Reims Collectables', '26.47.1555', "59 rue de l'Abbaye", '', 'Reims', '', '51100', 'France', 'EMEA', 'Henriot', 'Paul', 'Small'),
 ('10134', '41', '94.74', '2', '3884.34', '2003-07-01 00:00:00', 'Shipped', '3', '7', '2003', 'Motorcycles', '95', 'S10_1678', 'Lyon Souveniers', '+33 1 46 62 7555', '27 rue du Colonel Pierre Avia', '', 'Paris', '', '75508', 'France', 'EMEA', 'Da Cunha', 'Daniel', 'Medium'),
 ('10145', '45', '83.26', '6', '3746.7', '2003-08-25 00:00:00', 'Shipped', '3', '8', '2003', 'Motorcycles', '95', 'S10_1678', 'Toys4GrownUps.com', '6265557265', '78934 Hillside Dr.', '', 'Pasadena', 'CA', '90003', 'USA', 'NA', 'Young', 'Julie', 'Medium'),
 ('10159', '49', '100', '14', '5205.27', '2003-10-10 00:00:00', 'Shipped', '4', '10', '2003', 'Motorcycles', '95', 'S10_1678', 'Corporate Gift Ideas Co.', '6505551386', '7734 Strong St.', '', 'San Francisco', 'CA', '', 'USA', 'NA', 'Brown', 'Julie', 'Medium'),
 ('10168', '36', '96.66', '1', '3479.76', '2003-10-28 00:00:00', 'Shipped', '4', '10', '2003', 'Motorcycles', '95', 'S10_1678', 'Technics Stores Inc.', '6505556809', '9408 Furth Circle', '', 'Burlingame', 'CA', '94217', 'USA', 'NA', 'Hirano', 'Juri', 'Medium'),
 ('10180', '29', '86.13', '9', '2497.77', '2003-11-11 00:00:00', 'Shipped', '4', '11', '2003', 'Motorcycles', '95', 'S10_1678', 'Daedalus Designs Imports', '20.16.1555', '"184, chausse de Tournai"', '', 'Lille', '', '59000', 'France', 'EMEA', 'Rance', 'Martine', 'Small'),
 ('10188', '48', '100', '1', '5512.32', '2003-11-18 00:00:00', 'Shipped', '4', '11', '2003', 'Motorcycles', '95', 'S10_1678', 'Herkku Gifts', '+47 2267 3215', '"Drammen 121, PR 744 Sentrum"', '', 'Bergen', '', 'N 5804', 'Norway', 'EMEA', 'Oeztan', 'Veysel', 'Medium'),
 ('10201', '22', '98.57', '2', '2168.54', '2003-12-01 00:00:00', 'Shipped', '4', '12', '2003', 'Motorcycles', '95', 'S10_1678', 'Mini Wheels Co.', '6505555787', '5557 North Pendale Street', '', 'San Francisco', 'CA', '', 'USA', 'NA', 'Murphy', 'Julie', 'Small'),
 ('10211', '41', '100', '14', '4708.44', '2004-01-15 00:00:00', 'Shipped', '1', '1', '2004', 'Motorcycles', '95', 'S10_1678', 'Auto Canal Petit', '(1) 47.55.6555', '"25, rue Lauriston"', '', 'Paris', '', '75016', 'France', 'EMEA', 'Perrier', 'Dominique', 'Medium')]

Let's identify the unique values for some of the columns by using the DISTINCT function in SQL (for columns related to our questions of interest).

In [7]:
%sql SELECT DISTINCT(STATUS) FROM sales_data;

 * mysql+mysqlconnector://root:***@localhost/test_db
6 rows affected.


[('Shipped',),
 ('Disputed',),
 ('In Process',),
 ('Cancelled',),
 ('On Hold',),
 ('Resolved',)]

In [18]:
%sql SELECT DISTINCT(PRODUCTLINE) FROM sales_data;

 * mysql+mysqlconnector://root:***@localhost/test_db
7 rows affected.


[('Motorcycles',),
 ('Classic Cars',),
 ('Trucks and Buses',),
 ('Vintage Cars',),
 ('Planes',),
 ('Ships',),
 ('Trains',)]

In [13]:
%sql SELECT DISTINCT(YEAR_ID) FROM sales_data;

 * mysql+mysqlconnector://root:***@localhost/test_db
3 rows affected.


[('2003',), ('2004',), ('2005',)]

In [14]:
%sql SELECT DISTINCT(COUNTRY) FROM sales_data;

 * mysql+mysqlconnector://root:***@localhost/test_db
19 rows affected.


[('USA',),
 ('France',),
 ('Norway',),
 ('Australia',),
 ('Finland',),
 ('Austria',),
 ('UK',),
 ('Spain',),
 ('Sweden',),
 ('Singapore',),
 ('Canada',),
 ('Japan',),
 ('Italy',),
 ('Denmark',),
 ('Belgium',),
 ('Philippines',),
 ('Germany',),
 ('Switzerland',),
 ('Ireland',)]

In [16]:
%sql SELECT DISTINCT(CUSTOMERNAME) FROM sales_data;

 * mysql+mysqlconnector://root:***@localhost/test_db
92 rows affected.


[('Land of Toys Inc.',),
 ('Reims Collectables',),
 ('Lyon Souveniers',),
 ('Toys4GrownUps.com',),
 ('Corporate Gift Ideas Co.',),
 ('Technics Stores Inc.',),
 ('Daedalus Designs Imports',),
 ('Herkku Gifts',),
 ('Mini Wheels Co.',),
 ('Auto Canal Petit',),
 ('"Australian Collectors, Co."',),
 ('Vitachrome Inc.',),
 ('Tekni Collectables Inc.',),
 ('Gift Depot Inc.',),
 ('La Rochelle Gifts',),
 ("Marta's Replicas Co.",),
 ('"Toys of Finland, Co."',),
 ('Baane Mini Imports',),
 ('Diecast Classics Inc.',),
 ('Salzburg Collectables',),
 ('Souveniers And Things Co.',),
 ('FunGiftIdeas.com',),
 ('"UK Collectables, Ltd."',),
 ('Euro Shopping Channel',),
 ('"Volvo Model Replicas, Co"',),
 ('"Corrida Auto Replicas, Ltd"',),
 ('"Dragon Souveniers, Ltd."',),
 ('Classic Legends Inc.',),
 ('"Australian Gift Network, Co"',),
 ('"Classic Gift Ideas, Inc"',),
 ('"Saveley & Henriot, Co."',),
 ('Canadian Gift Exchange Network',),
 ('West Coast Collectables Co.',),
 ('Cambridge Collectables Co.',),
 ('Super Scale Inc.',),
 ('"Tokyo Collectables, Ltd"',),
 ('Amica Models & Co.',),
 ('Scandinavian Gift Ideas',),
 ('Auto Assoc. & Cie.',),
 ('Mini Gifts Distributors Ltd.',),
 ('Online Diecast Creations Co.',),
 ('Collectables For Less Inc.',),
 ('"Anna\'s Decorations, Ltd"',),
 ('Quebec Home Shopping Network',),
 ('Osaka Souveniers Co.',),
 ('Motor Mint Distributors Inc.',),
 ('Atelier graphique',),
 ('Mini Classics',),
 ('Danish Wholesale Imports',),
 ('"Stylish Desk Decors, Co."',),
 ('Mini Creations Ltd.',),
 ('Alpha Cognac',),
 ('"Double Decker Gift Stores, Ltd"',),
 ('Enaco Distributors',),
 ('"Men \'R\' US Retailers, Ltd."',),
 ('Collectable Mini Designs Co.',),
 ('Petit Auto',),
 ('"Royal Canadian Collectables, Ltd."',),
 ('Gifts4AllAges.com',),
 ('giftsbymail.co.uk',),
 ('"Oulu Toy Supplies, Inc."',),
 ('The Sharp Gifts Warehouse',),
 ('"La Corne D\'abondance, Co."',),
 ('Mini Auto Werke',),
 ('Diecast Collectables',),
 ('Handji Gifts& Co',),
 ('Cruz & Sons Co.',),
 ('Marseille Mini Autos',),
 ('"Toms Spezialitten, Ltd"',),
 ('"Vida Sport, Ltd"',),
 ("L'ordine Souveniers",),
 ('Online Mini Collectables',),
 ('"Blauer See Auto, Co."',),
 ('Suominen Souveniers',),
 ('Muscle Machine Inc',),
 ('"Clover Collections, Co."',),
 ('CAF Imports',),
 ('"AV Stores, Co."',),
 ('Heintze Collectables',),
 ('Microscale Inc.',),
 ('Boards & Toys Co.',),
 ('"Iberia Gift Imports, Corp."',),
 ('Signal Collectibles Ltd.',),
 ('Mini Caravy',),
 ('Signal Gift Stores',),
 ('"Norway Gifts By Mail, Co."',),
 ('Rovelli Gifts',),
 ('"Australian Collectables, Ltd"',),
 ('Gift Ideas Corp.',),
 ('"Bavarian Collectables Imports, Co."',),
 ('Royale Belge',),
 ('Auto-Moto Classics Inc.',)]

From the outputs above, it can be seen that the sales status was categorized into either “Shipped, Disputed, In Process, Cancelled, On Hold, and Resolved.” 
Data was collected over 3 years: 2003, 2004 and 2005. 
The company specializes in 7 product lines (Motorcycles, Classic Cars, Trucks and Buses, Vintage Cars, Planes, Ships, Trains), operates in 19 different countries and has a total of 92 customers.

Also, it appears there are no empty entries in the fields of our interest. So, let's proceed with the data exploration.

#### 1. Among all product lines, which one has the highest sales?
In order to find this, I group the sales by product line.

In [34]:
%%sql

SELECT 
    PRODUCTLINE,
    ROUND(SUM(SALES), 3) as Revenue
FROM 
    sales_data
GROUP BY 
    PRODUCTLINE
ORDER BY 
    Revenue DESC;


 * mysql+mysqlconnector://root:***@localhost/test_db
7 rows affected.


[('Classic Cars', 3919615.66),
 ('Vintage Cars', 1903150.84),
 ('Motorcycles', 1166388.34),
 ('Trucks and Buses', 1127789.84),
 ('Planes', 975003.57),
 ('Ships', 714437.13),
 ('Trains', 226243.47)]

Based on the displayed results, Classic Cars is the leading product line in terms of sales, while Trains exhibits the lowest sales among the product lines.

#### 2. Which year recorded the highest sales?
Now, group the data by the year.

In [26]:
%%sql

SELECT 
    YEAR_ID as Year,
    ROUND(SUM(SALES), 3) as Revenue
FROM 
    sales_data
GROUP BY 
    YEAR_ID
ORDER BY 
    Revenue DESC;


 * mysql+mysqlconnector://root:***@localhost/test_db
3 rows affected.


[('2004', 4724162.6), ('2003', 3516979.54), ('2005', 1791486.71)]

From the data presented above, it's evident that 2004 recorded the highest sales compared to the other two years. Interestingly, 2005 stands out with the lowest sales figure. This prompts the question: Why did the sales dip in 2005? So, let's explore that.

In [33]:
%%sql

SELECT 
    YEAR_ID as Year,
    MONTH_ID as Month
FROM 
    sales_data
GROUP BY 
    YEAR_ID, MONTH_ID
ORDER BY 
    Year ASC, CAST(Month AS UNSIGNED) ASC;

 * mysql+mysqlconnector://root:***@localhost/test_db
29 rows affected.


[('2003', '1'),
 ('2003', '2'),
 ('2003', '3'),
 ('2003', '4'),
 ('2003', '5'),
 ('2003', '6'),
 ('2003', '7'),
 ('2003', '8'),
 ('2003', '9'),
 ('2003', '10'),
 ('2003', '11'),
 ('2003', '12'),
 ('2004', '1'),
 ('2004', '2'),
 ('2004', '3'),
 ('2004', '4'),
 ('2004', '5'),
 ('2004', '6'),
 ('2004', '7'),
 ('2004', '8'),
 ('2004', '9'),
 ('2004', '10'),
 ('2004', '11'),
 ('2004', '12'),
 ('2005', '1'),
 ('2005', '2'),
 ('2005', '3'),
 ('2005', '4'),
 ('2005', '5')]

It appears that sales did not decline in 2005. Rather, sales data were only recorded for the first 5 months of 2005, in contrast to 2003 and 2004, which have sales data recorded for all 12 months.

#### 3. Which month experienced the highest sales?
So, to find out this, I iterate over the years 2003, 2004, and 2005. Then, the SQL query calculates the total sales revenue and the number of orders for each month in the specified year. The results are displayed below each year's heading.

In [115]:
for year in [2003, 2004, 2005]:
    display(f"Year: {year}")
    result = %sql SELECT \
                MONTH_ID as Month, \
                ROUND(SUM(SALES), 3) as Revenue, \
                COUNT(DISTINCT ORDERNUMBER) as Number_of_Orders \
            FROM \
                sales_data \
            WHERE \
                YEAR_ID = :year \
            GROUP BY \
                MONTH_ID \
            ORDER BY \
                Revenue DESC;
    display(result)


'Year: 2003'

 * mysql+mysqlconnector://root:***@localhost/test_db
12 rows affected.


[('11', 1029837.66, 28),
 ('10', 568290.97, 17),
 ('9', 263973.36, 8),
 ('12', 261876.46, 7),
 ('4', 201609.55, 7),
 ('8', 197809.3, 5),
 ('5', 192673.11, 6),
 ('7', 187731.88, 6),
 ('3', 174504.9, 6),
 ('6', 168082.56, 6),
 ('2', 140836.19, 3),
 ('1', 129753.6, 5)]

'Year: 2004'

 * mysql+mysqlconnector://root:***@localhost/test_db
12 rows affected.


[('11', 1089048.01, 33),
 ('10', 552924.25, 13),
 ('8', 461501.27, 12),
 ('12', 372802.66, 11),
 ('7', 327144.09, 10),
 ('9', 320750.91, 12),
 ('1', 316577.42, 8),
 ('2', 311419.53, 10),
 ('6', 286674.22, 10),
 ('5', 273438.39, 8),
 ('4', 206148.12, 10),
 ('3', 205733.73, 7)]

'Year: 2005'

 * mysql+mysqlconnector://root:***@localhost/test_db
5 rows affected.


[('5', 457861.06, 14),
 ('3', 374262.76, 12),
 ('2', 358186.18, 12),
 ('1', 339543.42, 12),
 ('4', 261633.29, 9)]

According to the results, November stands out as the month with the highest sales for both 2003 and 2004. Conversely, in 2005, May recorded the highest sales. However, it's important to note that the data for 2005 only spans five months.

In [56]:
for year in [2003, 2004]:
    display(f"Year: {year}")
    result = %sql \
        SELECT \
            MONTH_ID, \
            PRODUCTLINE, \
            SUM(SALES) as Revenue, \
            COUNT(ORDERNUMBER) as Number_of_Orders \
        FROM \
            sales_data \
        WHERE \
            YEAR_ID = :year and MONTH_ID = 11 \
        GROUP BY \
            MONTH_ID, PRODUCTLINE \
        ORDER BY \
            Revenue DESC;
    display(result)

'Year: 2003'

 * mysql+mysqlconnector://root:***@localhost/test_db
7 rows affected.


[('11', 'Classic Cars', 452924.3699999999, 114),
 ('11', 'Vintage Cars', 184673.39999999997, 66),
 ('11', 'Trucks and Buses', 127062.92000000003, 33),
 ('11', 'Motorcycles', 109345.5, 31),
 ('11', 'Ships', 79174.8, 27),
 ('11', 'Planes', 54133.270000000004, 16),
 ('11', 'Trains', 22523.4, 9)]

'Year: 2004'

 * mysql+mysqlconnector://root:***@localhost/test_db
7 rows affected.


[('11', 'Classic Cars', 372231.8899999999, 105),
 ('11', 'Vintage Cars', 233990.33999999997, 65),
 ('11', 'Motorcycles', 151711.85999999996, 39),
 ('11', 'Trucks and Buses', 123811.14000000001, 29),
 ('11', 'Planes', 121130.7, 36),
 ('11', 'Ships', 63900.85, 21),
 ('11', 'Trains', 22271.23, 6)]

Classic Cars is the product line that sold the most in November of 2003 and 2004. Trains, as in the question 1, remains the product line with the smallest sales.

#### 4. Which country achieved the highest product sales?

In [50]:
%%sql

SELECT 
    COUNTRY,
    ROUND(SUM(SALES), 3) as Revenue
FROM 
    sales_data
GROUP BY 
    COUNTRY
ORDER BY 
    Revenue DESC;

 * mysql+mysqlconnector://root:***@localhost/test_db
19 rows affected.


[('USA', 3627982.83),
 ('Spain', 1215686.92),
 ('France', 1110916.52),
 ('Australia', 630623.1),
 ('UK', 478880.46),
 ('Italy', 374674.31),
 ('Finland', 329581.91),
 ('Norway', 307463.7),
 ('Singapore', 288488.41),
 ('Denmark', 245637.15),
 ('Canada', 224078.56),
 ('Germany', 220472.09),
 ('Sweden', 210014.21),
 ('Austria', 202062.53),
 ('Japan', 188167.81),
 ('Switzerland', 117713.56),
 ('Belgium', 108412.62),
 ('Philippines', 94015.73),
 ('Ireland', 57756.43)]

The results show that the United States achieved the highest product sales, generating a revenue of $3,627,982.83. Following USA, Spain and France secured the second and third positions respectively in terms of sales revenue.

#### 5. Who stands out as the top customer?
To find this, I will perform RFM (Recency, Frequency, Monetary) analysis in SQL, and calculate the recency, frequency, and monetary value for each customer based on their transactions. I think these three key indicators are crucial in assessing customer behavior. Customers who have made purchases more recently are likely to be more engaged and valuable. Those who make frequent purchases are typically more loyal. Finally, clients who spend more money are considered higher value.

Based on the RFM scores, customers can be segmented into different categories to tailor marketing strategies. For example, customers with high RFM value are considered the most valuable and may be targeted with loyalty programs or exclusive offers. Customers with moderate scores may require nurturing to increase their engagement and spending. And those with low scores may need re-engagement strategies to encourage repeat purchases.

In [ ]:
%%sql

-- Convert string dates to date data types
UPDATE sales_data
SET ORDERDATE = STR_TO_DATE(ORDERDATE, '%m/%d/%Y %H:%i');

In [120]:
%%sql

-- Calculate Recency (R), Frequency (F), and Monetary Value (M) for each customer
WITH customer_rfm AS (
    SELECT
        CUSTOMERNAME as CustomerName,
        MAX(ORDERDATE) as LastPurchaseDate,
        COUNT(ORDERNUMBER) as NumberOfTransactions,
        SUM(SALES) AS MoneyValue
    FROM
        sales_data
    GROUP BY
        CUSTOMERNAME
)

-- Calculate Recency
SELECT
    CustomerName,
    LastPurchaseDate,
    DATEDIFF(CURRENT_DATE, LastPurchaseDate) as Recency,
    NumberOfTransactions,
    MoneyValue
FROM
    customer_rfm
ORDER BY
    MoneyValue DESC
LIMIT 10; 


 * mysql+mysqlconnector://root:***@localhost/test_db
10 rows affected.


[('Euro Shopping Channel', '2005-05-31 00:00:00', 6896, 259, 912294.1100000002),
 ('Mini Gifts Distributors Ltd.', '2005-05-29 00:00:00', 6898, 180, 654858.06),
 ('"Australian Collectors, Co."', '2004-11-29 00:00:00', 7079, 55, 200995.40999999997),
 ('Muscle Machine Inc', '2004-12-01 00:00:00', 7077, 48, 197736.93999999997),
 ('La Rochelle Gifts', '2005-05-31 00:00:00', 6896, 53, 180124.9),
 ('"Dragon Souveniers, Ltd."', '2005-03-02 00:00:00', 6986, 43, 172989.68000000008),
 ('Land of Toys Inc.', '2004-11-15 00:00:00', 7093, 49, 164069.44000000003),
 ('The Sharp Gifts Warehouse', '2005-04-22 00:00:00', 6935, 40, 160010.26999999996),
 ('"AV Stores, Co."', '2004-11-17 00:00:00', 7091, 51, 157807.80999999997),
 ('"Anna\'s Decorations, Ltd"', '2005-03-09 00:00:00', 6979, 46, 153996.13000000003)]

The recency column shows the number of days since each customer's last purchase. Customers like "Euro Shopping Channel" and "Mini Gifts Distributors Ltd." made purchases relatively recently, with recency values of 6895 and 6897 days, respectively. On the other hand, customers like "Boards & Toys Co." and "Atelier graphique" made purchases quite a while ago, with recency values of 7007 and 7082 days, respectively.

The frequency column represents the number of transactions made by each customer. Customers such as "Euro Shopping Channel" and "Mini Gifts Distributors Ltd." have high frequencies, indicating they made numerous purchases. Conversely, customers like "Bavarian Collectables Imports, Co." and "Boards & Toys Co." have lower frequencies, suggesting they made fewer purchases.

Customers such as "Euro Shopping Channel" and "Mini Gifts Distributors Ltd." have made significant monetary contributions. Meanwhile, customers like "Boards & Toys Co." and "Atelier graphique" have lower total spending.

Overall, high-value customers who made purchases recently, frequently, and spent a considerable amount are "Euro Shopping Channel" and "Mini Gifts Distributors Ltd." 
Customers with relatively low recency, frequency, and monetary value might require targeted marketing efforts or incentives to increase their engagement and spending.
Customers who made purchases quite a while ago and have low spending might be considered inactive or lost customers, requiring different retention strategies.

####

To gain a deeper understanding of the customers, I wanted to categorize customers into distinct segments such as 'new customers', 'potential churners', 'active', 'loyal', and 'slipping away, cannot lose'. This segmentation provides valuable insights that can inform strategies for the sales and marketing teams, helping them better target and engage with different customer groups.

- 'lost_customers': These are customers who have low recency, frequency, and monetary value. They used to be active customers but have stopped purchasing from the company.

- 'slipping_away': These are customers who were previously active but are now showing a decline in recency, frequency, and monetary value. They are at risk of becoming lost customers if their behavior does not change.

- 'new_customers': These are newly acquired customers who have recently made their first purchase. They are still in the early stages of their relationship with the company.

- 'potential_churners': These are customers who are showing signs of churn based on their RFM scores. Although they may still be active, their behavior suggests that they could become lost customers in the future if certain actions are not taken.

- 'active': These are active and engaged customers who have high recency, frequency, and monetary value. They are valuable to the business and are likely to continue making purchases.

- 'loyal': These are loyal customers who consistently show high recency, frequency, and monetary value. They are the most valuable segment as they contribute significantly to the company's revenue and are likely to remain customers for the long term.

In [119]:
%%sql

-- Calculate RFM (Recency, Frequency, Monetary Value) metrics for each customer
WITH rfm as (
    SELECT 
        CUSTOMERNAME as CustomerName,
        ROUND(SUM(SALES), 3) as MonetaryValue,
        ROUND(AVG(SALES), 3) as AvgMonetaryValue,
        COUNT(ORDERNUMBER) as Frequency,
        MAX(ORDERDATE) as LastOrderDate,
        (SELECT MAX(ORDERDATE) FROM sales_data) as MaxOrderDate,
        DATEDIFF(CURRENT_DATE, MAX(ORDERDATE)) as Recency
    FROM 
        sales_data
    GROUP BY 
        CUSTOMERNAME
),
rfm_calc as (
    -- Calculate quartile ranks for Recency, Frequency, and MonetaryValue
    SELECT 
        r.*,
        NTILE(4) OVER (ORDER BY Recency DESC) as rfm_recency,
        NTILE(4) OVER (ORDER BY Frequency) as rfm_frequency,
        NTILE(4) OVER (ORDER BY MonetaryValue) as rfm_monetary
    FROM 
        rfm r
)

-- Final selection and categorization of customers based on RFM quartiles.  
-- This part calculates the RFM cell value for each customer by summing up their recency, frequency, and monetary quartiles. 
SELECT 
    c.*, 
    rfm_recency + rfm_frequency + rfm_monetary as rfm_cell,
    CONCAT(rfm_recency, rfm_frequency, rfm_monetary) as rfm_cell_string,
    CASE 
        WHEN CONCAT(rfm_recency, rfm_frequency, rfm_monetary) IN ('111', '112', '121', '122', '123', '132', '211', '212', '114', '141') THEN 'lost_customers'
        WHEN CONCAT(rfm_recency, rfm_frequency, rfm_monetary) IN ('133', '134', '143', '244', '334', '343', '344', '144') THEN 'slipping_away'
        WHEN CONCAT(rfm_recency, rfm_frequency, rfm_monetary) IN ('311', '411', '331') THEN 'new_customers'
        WHEN CONCAT(rfm_recency, rfm_frequency, rfm_monetary) IN ('222', '223', '233', '322') THEN 'potential_churners'
        WHEN CONCAT(rfm_recency, rfm_frequency, rfm_monetary) IN ('323', '333', '321', '422', '332', '432') THEN 'active'
        WHEN CONCAT(rfm_recency, rfm_frequency, rfm_monetary) IN ('433', '434', '443', '444') THEN 'loyal'
    END as rfm_segment
FROM 
    rfm_calc c
ORDER BY
    MonetaryValue DESC
LIMIT 10; 



 * mysql+mysqlconnector://root:***@localhost/test_db
10 rows affected.


[('Euro Shopping Channel', 912294.11, 3522.371, 259, '2005-05-31 00:00:00', '2005-05-31 00:00:00', 6896, 4, 4, 4, 12, '444', 'loyal'),
 ('Mini Gifts Distributors Ltd.', 654858.06, 3638.1, 180, '2005-05-29 00:00:00', '2005-05-31 00:00:00', 6898, 4, 4, 4, 12, '444', 'loyal'),
 ('"Australian Collectors, Co."', 200995.41, 3654.462, 55, '2004-11-29 00:00:00', '2005-05-31 00:00:00', 7079, 3, 4, 4, 11, '344', 'slipping_away'),
 ('Muscle Machine Inc', 197736.94, 4119.52, 48, '2004-12-01 00:00:00', '2005-05-31 00:00:00', 7077, 3, 4, 4, 11, '344', 'slipping_away'),
 ('La Rochelle Gifts', 180124.9, 3398.583, 53, '2005-05-31 00:00:00', '2005-05-31 00:00:00', 6896, 4, 4, 4, 12, '444', 'loyal'),
 ('"Dragon Souveniers, Ltd."', 172989.68, 4023.016, 43, '2005-03-02 00:00:00', '2005-05-31 00:00:00', 6986, 3, 4, 4, 11, '344', 'slipping_away'),
 ('Land of Toys Inc.', 164069.44, 3348.356, 49, '2004-11-15 00:00:00', '2005-05-31 00:00:00', 7093, 2, 4, 4, 10, '244', 'slipping_away'),
 ('The Sharp Gifts Warehouse', 160010.27, 4000.257, 40, '2005-04-22 00:00:00', '2005-05-31 00:00:00', 6935, 4, 4, 4, 12, '444', 'loyal'),
 ('"AV Stores, Co."', 157807.81, 3094.271, 51, '2004-11-17 00:00:00', '2005-05-31 00:00:00', 7091, 2, 4, 4, 10, '244', 'slipping_away'),
 ('"Anna\'s Decorations, Ltd"', 153996.13, 3347.742, 46, '2005-03-09 00:00:00', '2005-05-31 00:00:00', 6979, 3, 4, 4, 11, '344', 'slipping_away')]

Let's find all loyal customers. They are the most valuable segment as they contribute significantly to the company's revenue and are likely to remain customers for the long term.

In [114]:
%%sql

-- Calculate RFM (Recency, Frequency, Monetary Value) metrics for each customer
WITH rfm as (
    SELECT 
        CUSTOMERNAME as CustomerName,
        ROUND(SUM(SALES), 3) as MonetaryValue,
        ROUND(AVG(SALES), 3) as AvgMonetaryValue,
        COUNT(ORDERNUMBER) as Frequency,
        MAX(ORDERDATE) as LastOrderDate,
        (SELECT MAX(ORDERDATE) FROM sales_data) as MaxOrderDate,
        DATEDIFF(CURRENT_DATE, MAX(ORDERDATE)) as Recency
    FROM 
        sales_data
    GROUP BY 
        CUSTOMERNAME
),
rfm_calc as (
    -- Calculate quartile ranks for Recency, Frequency, and MonetaryValue
    SELECT 
        r.*,
        NTILE(4) OVER (ORDER BY Recency DESC) as rfm_recency,
        NTILE(4) OVER (ORDER BY Frequency) as rfm_frequency,
        NTILE(4) OVER (ORDER BY MonetaryValue) as rfm_monetary
    FROM 
        rfm r
)

-- Final selection and categorization of customers based on RFM quartiles.  
-- This part calculates the RFM cell value for each customer by summing up their recency, frequency, and monetary quartiles. 
SELECT 
    c.*, 
    rfm_recency + rfm_frequency + rfm_monetary as rfm_cell,
    CONCAT(rfm_recency, rfm_frequency, rfm_monetary) as rfm_cell_string,
    CASE 
        WHEN CONCAT(rfm_recency, rfm_frequency, rfm_monetary) IN ('111', '112', '121', '122', '123', '132', '211', '212', '114', '141') THEN 'lost_customers'
        WHEN CONCAT(rfm_recency, rfm_frequency, rfm_monetary) IN ('133', '134', '143', '244', '334', '343', '344', '144') THEN 'slipping_away'
        WHEN CONCAT(rfm_recency, rfm_frequency, rfm_monetary) IN ('311', '411', '331') THEN 'new_customers'
        WHEN CONCAT(rfm_recency, rfm_frequency, rfm_monetary) IN ('222', '223', '233', '322') THEN 'potential_churners'
        WHEN CONCAT(rfm_recency, rfm_frequency, rfm_monetary) IN ('323', '333', '321', '422', '332', '432') THEN 'active'
        WHEN CONCAT(rfm_recency, rfm_frequency, rfm_monetary) IN ('433', '434', '443', '444') THEN 'loyal'
    END as rfm_segment
FROM 
    rfm_calc c
HAVING
    rfm_segment = 'loyal'
ORDER BY
    MonetaryValue DESC;  


 * mysql+mysqlconnector://root:***@localhost/test_db
14 rows affected.


[('Euro Shopping Channel', 912294.11, 3522.371, 259, '2005-05-31 00:00:00', '2005-05-31 00:00:00', 6895, 4, 4, 4, 12, '444', 'loyal'),
 ('Mini Gifts Distributors Ltd.', 654858.06, 3638.1, 180, '2005-05-29 00:00:00', '2005-05-31 00:00:00', 6897, 4, 4, 4, 12, '444', 'loyal'),
 ('La Rochelle Gifts', 180124.9, 3398.583, 53, '2005-05-31 00:00:00', '2005-05-31 00:00:00', 6895, 4, 4, 4, 12, '444', 'loyal'),
 ('The Sharp Gifts Warehouse', 160010.27, 4000.257, 40, '2005-04-22 00:00:00', '2005-05-31 00:00:00', 6934, 4, 4, 4, 12, '444', 'loyal'),
 ('Souveniers And Things Co.', 151570.98, 3295.021, 46, '2005-05-29 00:00:00', '2005-05-31 00:00:00', 6897, 4, 4, 4, 12, '444', 'loyal'),
 ('Salzburg Collectables', 149798.63, 3744.966, 40, '2005-05-17 00:00:00', '2005-05-31 00:00:00', 6909, 4, 4, 4, 12, '444', 'loyal'),
 ('Danish Wholesale Imports', 145041.6, 4028.933, 36, '2005-04-15 00:00:00', '2005-05-31 00:00:00', 6941, 4, 4, 4, 12, '444', 'loyal'),
 ("L'ordine Souveniers", 142601.33, 3656.444, 39, '2005-05-10 00:00:00', '2005-05-31 00:00:00', 6916, 4, 4, 4, 12, '444', 'loyal'),
 ('Reims Collectables', 135042.94, 3293.73, 41, '2005-03-30 00:00:00', '2005-05-31 00:00:00', 6957, 4, 4, 4, 12, '444', 'loyal'),
 ('Diecast Classics Inc.', 122138.14, 3939.94, 31, '2005-05-30 00:00:00', '2005-05-31 00:00:00', 6896, 4, 3, 4, 11, '434', 'loyal'),
 ('"Tokyo Collectables, Ltd"', 120562.74, 3767.586, 32, '2005-04-22 00:00:00', '2005-05-31 00:00:00', 6934, 4, 3, 3, 10, '433', 'loyal'),
 ('"UK Collectables, Ltd."', 118008.27, 4069.251, 29, '2005-04-08 00:00:00', '2005-05-31 00:00:00', 6948, 4, 3, 3, 10, '433', 'loyal'),
 ('Handji Gifts& Co', 115498.73, 3208.298, 36, '2005-04-23 00:00:00', '2005-05-31 00:00:00', 6933, 4, 4, 3, 11, '443', 'loyal'),
 ('Auto Canal Petit', 93170.66, 3450.765, 27, '2005-04-07 00:00:00', '2005-05-31 00:00:00', 6949, 4, 3, 3, 10, '433', 'loyal')]

#### 6. What products are usually sold together?

In [118]:
%%sql

WITH TransactionItems AS (
    -- Identifying transaction items (products) for each order along with product line
    SELECT s.ORDERNUMBER, s.PRODUCTCODE, s.PRODUCTLINE
    FROM sales_data s
),
OrderSummary AS (
    -- Aggregating transaction data to summarize products purchased together in each order
    SELECT ORDERNUMBER, GROUP_CONCAT(PRODUCTCODE ORDER BY PRODUCTCODE) AS ProductList
    FROM TransactionItems
    GROUP BY ORDERNUMBER
),
ProductPairs AS (
    -- Generating all possible pairs of products within each order along with product lines
    SELECT o1.ORDERNUMBER, p1.PRODUCTCODE AS Product1, p1.PRODUCTLINE AS ProductLine1,
           p2.PRODUCTCODE AS Product2, p2.PRODUCTLINE AS ProductLine2
    FROM OrderSummary o1
    JOIN TransactionItems p1 ON o1.ORDERNUMBER = p1.ORDERNUMBER
    JOIN TransactionItems p2 ON o1.ORDERNUMBER = p2.ORDERNUMBER
    WHERE p1.PRODUCTCODE < p2.PRODUCTCODE -- Ensure unique pairs
)
-- Counting occurrences of each product pair across all orders
SELECT Product1, ProductLine1, Product2, ProductLine2, COUNT(*) AS PairFrequency
FROM ProductPairs
GROUP BY Product1, ProductLine1, Product2, ProductLine2
ORDER BY PairFrequency DESC
LIMIT 10; 

 * mysql+mysqlconnector://root:***@localhost/test_db
10 rows affected.


[('S700_2047', 'Ships', 'S72_1253', 'Planes', 26),
 ('S24_3949', 'Planes', 'S700_4002', 'Planes', 26),
 ('S18_2319', 'Trucks and Buses', 'S18_3232', 'Classic Cars', 26),
 ('S24_2841', 'Planes', 'S24_3420', 'Vintage Cars', 26),
 ('S50_1341', 'Vintage Cars', 'S700_1691', 'Planes', 26),
 ('S18_2957', 'Vintage Cars', 'S18_3136', 'Vintage Cars', 26),
 ('S10_4962', 'Classic Cars', 'S18_4600', 'Trucks and Buses', 25),
 ('S700_3962', 'Ships', 'S72_3212', 'Ships', 25),
 ('S10_1949', 'Classic Cars', 'S18_1097', 'Trucks and Buses', 25),
 ('S18_1129', 'Classic Cars', 'S18_1984', 'Classic Cars', 25)]

Knowing which products tend to be bought together is very important. It helps to figure out what customers really like to buy together, so the company can put those items side by side or suggest them as a package deal. This not only makes shopping easier for customers but also boosts sales by encouraging customers to buy more. By understanding these buying patterns, companies can create targeted promotions and marketing campaigns that resonate with customers' preferences, making their shopping experience even better and keeping them coming back for more.